analyzers/src/SonarAnalyzer.CSharp/Common/Walkers/CatchLoggingInvocationWalker.cs

```diff
@@ -0,0 +1,119 @@
+    public ThrowStatementSyntax ThrowStatementSyntax { get; private set; }
+    public ImmutableArray<InvocationExpressionSyntax> LoggingInvocationsWithoutException { get; private set; } = ImmutableArray<InvocationExpressionSyntax>.Empty;
+
+    private static readonly ImmutableArray<LoggingInvocationDescriptor> LoggingInvocationIdentifiers = ImmutableArray.Create(
+        new LoggingInvocationDescriptor(MicrosoftExtensionsLogging, KnownType.Microsoft_Extensions_Logging_LoggerExtensions, false),
+        new LoggingInvocationDescriptor(CastleCoreOrCommonCore, KnownType.Castle_Core_Logging_ILogger, true),
+        new LoggingInvocationDescriptor(CastleCoreOrCommonCore, KnownType.Common_Logging_ILog, true),
+        new LoggingInvocationDescriptor(Log4NetILog, KnownType.log4net_ILog, true),
+        new LoggingInvocationDescriptor(Log4NetILogExtensions, KnownType.log4net_Util_ILogExtensions, false),
+        new LoggingInvocationDescriptor(NLogLoggingMethods, KnownType.NLog_ILogger, true),
+        new LoggingInvocationDescriptor(NLogLoggingMethods, KnownType.NLog_ILoggerExtensions, false),
+        new LoggingInvocationDescriptor(NLogILoggerBase, KnownType.NLog_ILoggerBase, true),
+        new LoggingInvocationDescriptor(Serilog, KnownType.Serilog_ILogger, true),
+        new LoggingInvocationDescriptor(Serilog, KnownType.Serilog_Log, false));
+
+    public CatchLoggingInvocationWalker(SemanticModel model)
+    {
+        this.model = model;
+    }
+
+    public override void VisitCatchClause(CatchClauseSyntax node)
+    {
+        // We want to look for logging invocations only in the main catch clause.
+        if (isFirstCatchClauseVisited)
+        {
+            return;
+        }
+
+        isFirstCatchClauseVisited = true;
+        hasWhenFilterWithDeclarations = node.Filter != null && node.Filter.DescendantNodes().Any(DeclarationPatternSyntaxWrapper.IsInstance);
+        if (node.Declaration != null && !node.Declaration.Identifier.IsKind(SyntaxKind.None))
+        {
+            CaughtException = model.GetDeclaredSymbol(node.Declaration);
+        }
+        base.VisitCatchClause(node);
+    }
+
+    public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+    {
+        if (!IsExceptionLogged && IsLoggingInvocation(node, model))
+        {
+            if (GetArgumentSymbolDerivedFromException(node, model) is { } currentException
+                && (hasWhenFilterWithDeclarations || currentException.Equals(CaughtException)))
+            {
+                IsExceptionLogged = true;
+                LoggingInvocationWithException = node;
+                return;
+            }
+            else
+            {
+                LoggingInvocationsWithoutException = LoggingInvocationsWithoutException.Add(node);
+            }
+        }
+        base.VisitInvocationExpression(node);
+    }
+
+    public override void VisitThrowStatement(ThrowStatementSyntax node)
+    {
+        ThrowStatementSyntax = node;
```

Comment:
There is a small issue with this line: If there are multiple throw statements, this is essentially a `.Last()`.

For example:

```csharp
    public void Overwrite(ILogger logger)
    {
        try
        { }
        catch (Exception baseball) // Compliant FN
        {
            logger.LogError(baseball, "Message");
            throw baseball;                     // ThrowStatementSyntax set to this

            throw new Exception("Outplayed");   // ThrowStatementSyntax re-set to this
        }
    }
```

IMO it makes more sense to semantically do `.First()`, so I would do something like:
```csharp
ThrowStatementSyntax ??= node;
```